REVIEW: Helper objects and methods for canvas resources

nein09 opened the pull request:
For https://github.com/sul-dlss/iiifManifestLayouts/issues/40

nein09 committed:
Initial sketch of what the canvas helper might look like

nein09 committed:
Fix tabs

nein09 committed:
Rename canvasImageState to canvasObjects

nein09 committed:
Rename canvasState to viewerState for clarity.

iangilman commented on src/canvasObject.js:
```diff
@@ -0,0 +1,18 @@
+'use strict';
+
+var canvasObject= function(canvas) {
```
I imagine eventually we'll want this to be a robust object with methods of its own. For that I recommend something like this: 

```
'use strict';

var CanvasObject = function(config) {
    this.needed = false;
    this.visible = true;
    this.clipRegion = null;
    this.opacity = 1;
    this.position = {
        x: 0,
        y: 0
    };
    this.placeholder = '/some-image.jpg';
    this.tileSourceUrl = config.canvas.images[0].resource.service['@id'] + '/info.json';
};

CanvasObject.prototype = {
   myMethod: function() {
   } 
};

module.exports = CanvasObject;
```

...which would be created in main.js like so: 

```
 canvasObjects[canvas['@id']] = CanvasObject({
   canvas: canvas
 });
```

nein09 committed:
Don't clobber existing attributes in the canvas object

iangilman commented on src/canvasObject.js:
```diff
@@ -0,0 +1,21 @@
+'use strict';
+
+var _ = require('underscore');
```
@aeschylus Are we ok with having underscore be a dependency for the built iiifManifestLayouts module? I'm all for it; it'll definitely simplify things. Just making sure.

iangilman commented:
@nein09 Looks like we've been using two space indents for this project (which, incidentally, is what I prefer). Please change canvasObject.js to match. And I suppose canvasLayout.js while you're at it.

nein09 committed:
Fix tabs

nein09 committed:
Make CanvasObject into a first-class object

nein09 committed:
Make CanvasObject set its own main image object

iangilman commented on src/canvasObject.js:
```diff
@@ -0,0 +1,27 @@
+CanvasObject.prototype = {
+   setMainImage: function(mainImage) {
+    this.mainImageObj = mainImage;
+    // this object also has things like opacity - should we make sure that corresponding values of this
```
I imagine we should make sure the mainImage conforms to the settings of the CanvasObject.

iangilman commented:
FWIW, I'm all for using Underscore...I just want to make sure @aeschylus is cool with it.

nein09 commented:
Don't worry- I took underscore out that time because I wasn't using it anymore - it was nice when I was just merging keys, but I wanted to be more explicit about building up that object.

mejackreed commented:
Prefer not to include another library if not necessary.

nein09 commented:
That makes sense- I initially included it because it's in package.json anyway, but I see now that it's part of the dependency chain here, not explicitly chosen as something include.

If I really feel like I want it for something, we can talk about this, but for now, I'm fine with leaving it out.

iangilman committed:
Added sublime project

iangilman commented:
@nein09 I've added a sublime project to this branch; if you use it, it can help you keep your tab settings in line. Also note that it's excluding the node_modules from the file list; this makes multi-file searching more pleasant. We might consider also excluding build results.

nein09 committed:
Move fade function into a helper file

nein09 committed:
Add tiled image to viewer inside canvasObject

nein09 committed:
Add *.DS_Store to .gitignore

nein09 committed:
Add newline at EOF

nein09 committed:
Use event object correctly

nein09 committed:
Upgrade to OpenSeadragon 2.1.0

nein09 committed:
Use a thumbnail for default images

nein09 committed:
Change from thumbnail to full tilesource on canvas click.

nein09 commented:
Okay, I think I've got this in a good place now. I added some properties and methods to canvasObject to make it more of a full-fledged object - @aeschylus , is there anything else you'd like to see in there in order to consider https://github.com/sul-dlss/iiifManifestLayouts/issues/40 resolved?

Currently, the viewer loads thumbnails initially, and doesn't load the actual tileSource until you click on the image. We should talk about when else to load the actual tileSource - knowing that you're in detail mode on a selected canvas seems like a good time. When that happens, I think I'd want to change to a more fine-grained event system, as we discussed last week.

Please note that I'm using the same dummy thumbnail for each image right now- we should talk about how to move forward with the real ones (does IIIF have a thumbnail attribute in the manifest? If so, what do we do when it's missing?).

iangilman commented:
We'll also want to load the actual tileSource if you're in filmstrip mode and you pan over to it. 

One approach would be to watch zoom and pan events and check to see if a tiledImage is featured enough to swap. That way we don't need to pay attention to specific scenarios that might feature a canvas.

iangilman commented on src/canvasObject.js:
```diff
@@ -0,0 +1,82 @@
+};
+
+CanvasObject.prototype = {
+  openTileSource: function(x, y, width, viewer) {
```
This should just be a swap, so it should already know the x, y, width, viewer, instead of having it passed in. 

Also, it should know whether the swap has already occurred (i.e. it's already a full tile source) and just return without doing an additional swap.

iangilman commented on src/canvasObject.js:
```diff
@@ -0,0 +1,82 @@
+    });
+  },
+
+  openThumbnail: function(x, y, width, viewer) {
```
For that matter, maybe the x, y, width, viewer needed here should already have been passed in during the object constructor? 

The goal is for these objects to be as much in charge of their own destiny as possible, not relying on outside source of data (except at creation of course). 

On a related note, when we get into having alternate images and overlays, we'll probably want the CanvasObject to have its own idea of its bounds, independent of any of its tiled images. Kind of like how I did it here: 

https://github.com/openseadragon/openseadragon/blob/master/test/demo/m2/js/page.js

iangilman commented on src/canvasAnimationHelpers.js:
```diff
@@ -0,0 +1,27 @@
+'use strict';
+
+var canvasAnimationHelpers = {
+ fade: function(image, targetOpacity, callback) {
```
Seems like this should just be part of CanvasObject rather than being its own module.

nein09 committed:
Remove DS_Store from gitignore

nein09 committed:
Put fade in canvasObject

nein09 commented:
Thanks! I don't mind where DS_Store gets ignored; it's easy enough for me to make that global.

nein09 commented:
Passing the position to the constructor means changing where/when the canvasObjects get constructed - right now, they're built before a layout is calculated. I could make the objects be constructed on the fly, in enterImages, though.

I also added some helpers to canvasObject so that we won't have to reference mainImageObj directly (like we do now in translateTileSources) - I think it'd be nicer to make mainImageObj a private member of CanvasObject. What do you all think?

nein09 committed:
Make CanvasObject have a better grasp on its own bounds and position, and use them

iangilman commented:
I like the latest commit...you can pass the x and y into the constructor or call `setPosition` later...that gives the most flexibility as to when the canvasObjects get created vs. when their location is determined. 

And yes, I'm in favor of mainImageObj being private to CanvasObject; that should drive some good encapsulation.

iangilman commented on src/canvasObject.js:
```diff
@@ -0,0 +1,143 @@
+      return;
+    }
+    var self = this;
+    var bounds = self.mainImageObj.getBounds();
```
I don't know that I have good reasons to back this up, but my preference is to not use `self` in places where you don't need it, e.g. this line. My preference would be for this line to use `this` and the code inside the success function to use `self` (where it's necessary). 

Also, I guess just a stylistic thing, but I like having the `var self = this` line always at the top of the function if it's present...that way you don't have to go hunting for it. 

More of my `self` thoughts: http://blog.iangilman.com/2015/02/self-this.html

iangilman commented on src/canvasObject.js:
```diff
@@ -0,0 +1,143 @@
+    }
+    var self = this;
+    var bounds = self.mainImageObj.getBounds();
+    viewer.viewport.fitBounds(bounds);
```
This `fitBounds` is just here because currently `openTileSource` is only called when the user clicks on the image...in the future we'll probably call it in other contexts, so the `fitBounds` should probably come out of here. 

Also, CanvasObject probably shouldn't be in charge of moving the viewport around at all...it should just keep track of its own self.

iangilman commented on src/canvasObject.js:
```diff
@@ -0,0 +1,143 @@
+
+  // Call this to make sure that the CanvasObject's information about the world is the same as the OSD image's.
+  _setBoundsInternal: function() {
+    if(this.mainImageObj) {
```
I think ultimately we'll want the data flow to always be from the CanvasObject's bounds to the TiledImage; never in reverse. That way it's clear who's running the show.

nein09 committed:
Set bounds in click handler but not openTileSource

nein09 committed:
Encapsulate OSD image object in CanvasObject

nein09 committed:
Put CanvasObject in charge of position and dimensions

nein09 committed:
Set the height too

nein09 committed:
Open the tileSource when a canvas is selected

iangilman commented on src/main.js:
```diff
@@ -319,74 +315,37 @@ var manifestor = function(options) {
-    OpenSeadragon.requestAnimationFrame(frame);
-  };
+    canvasImageState.setPosition(canvasData.x, canvasData.y);
+    canvasImageState.setHeight(canvasData.height);
```
One thing to keep in mind is that OSD TiledImages don't support stretching their aspect ratio, so internally they just have a single scale value. This means when you set the height it changes the scale and then when you set the width it changes the scale again. So, at the OSD level it's more efficient to just set one or the other. 

On the other hand, the CanvasObject probably could use both width and height. So maybe having a `setSize` method, that keeps track of both width and height but internally just tells the TiledImage to set its width, would be good.

iangilman commented:
Beyond that minor comment, it's looking great to me! The other thing that still needs doing is getting the actual thumbnails in there, which hopefully @aeschylus can fill us in on.

nein09 committed:
Use setSize on CanvasObject
Only set the width on OSD objects, because they get their height from their aspect ratio.

nein09 committed:
Use canvasObjects for previous/next behavior

This includes a refactor to use CanvasObject for as much information as possible. This change started as a fix for the loading the tilesources for both pages in a paired page view, and it grew.

iangilman commented:
Changes look good to me.

iangilman committed:
Example: resize events before viewer created were causing exceptions

iangilman committed:
Example: fixed styling for Firefox

iangilman commented:
I've made a couple of commits...just little browser compatibility things.

nein09 committed:
Open tilesource directly instead of thumbnail

nein09 committed:
Set image object values when setting as main image obj

iangilman commented:
Changes look good to me. I think this is ready to merge. Can you please confirm, @aeschylus?
